vault/identity_store.go

```diff
@@ -861,110 +875,186 @@ func (i *IdentityStore) invalidateClientPath(ctx context.Context, key string) {
-			i.logger.Error("failed to delete entity from MemDB", "entity_id", entity.ID, "error", err)
-			return
-		}
-	}
-
-	// Now read the invalidated storage key
+	// Get local aliases from the invalidated bucket
 	bucket, err := i.localAliasPacker.GetBucket(ctx, key)
 	if err != nil {
 		i.logger.Error("failed to refresh local aliases", "key", key, "error", err)
 		return
 	}
+
+	// This map tracks the set of local aliases that need to be updated in each
+	// affected entity in MemDB.
+	entityLocalAliasesToUpsert := map[*identity.Entity][]*identity.Alias{}
+
+	// This map tracks the set of local aliases that need to be removed from
+	// their affected entity in MemDB, as well as removing the local alias
+	// themselves.
+	entityLocalAliasesToRemove := map[*identity.Entity][]*identity.Alias{}
+
 	if bucket != nil {
+		// The storage entry for the local alias bucket exists, so we need to
+		// compare the local aliases in that bucket with those in MemDB and only
+		// update those that are different. The local aliases in the bucket are
+		// the source of truth.
+
+		// Iterate over each local alias item from the bucket
 		for _, item := range bucket.Items {
 			if strings.HasSuffix(item.ID, tmpSuffix) {
 				continue
 			}
 
-			var localAliases identity.LocalAliases
-			err = ptypes.UnmarshalAny(item.Message, &localAliases)
+			var bucketLocalAliases identity.LocalAliases
+
+			err = anypb.UnmarshalTo(item.Message, &bucketLocalAliases, proto.UnmarshalOptions{})
 			if err != nil {
-				i.logger.Error("failed to parse local aliases during invalidation", "error", err)
+				i.logger.Error("failed to parse local aliases during invalidation", "item_id", item.ID, "error", err)
 				return
 			}
-			for _, alias := range localAliases.Aliases {
-				// Add to the aliases table
-				if err := i.MemDBUpsertAliasInTxn(txn, alias, false); err != nil {
-					i.logger.Error("failed to insert local alias to memdb during invalidation", "error", err)
+
+			for _, bucketLocalAlias := range bucketLocalAliases.Aliases {
+				// Find the entity related to bucketLocalAlias in MemDB in order
+				// to track any local aliases modifications that must be made in
+				// this entity.
+				memDBEntity := i.FetchEntityForLocalAliasInTxn(txn, bucketLocalAlias)
+				if memDBEntity == nil {
+					// FetchEntityForLocalAliasInTxn already logs any error
 					return
 				}
 
-				// Fetch the associated entity and add the alias to that too.
-				entity, err := i.MemDBEntityByIDInTxn(txn, alias.CanonicalID, false)
-				if err != nil {
-					i.logger.Error("failed to fetch entity during local alias invalidation", "error", err)
-					return
+				// memDBLocalAlias starts off nil but gets set to the local
+				// alias from memDBLocalAliases whose ID matches the ID of
+				// bucketLocalAlias.
+				var memDBLocalAlias *identity.Alias
+				for i, localAlias := range memDBLocalAliases {
+					if localAlias.ID == bucketLocalAlias.ID {
```

Comment:
Discussed out of band about the possibility of creating a new `MemDBMapIDToEntityByBucketKeyInTxn` (yuck). It would be really nice if we could build up a `map[ID]identity.Entity` and `map[ID]identity.LocalAlias` and use those to trim duplicates by direct lookup, thus replacing the inner loop.

I think this could potentially be a significant improvement in clarity and performance (though the current patch has already proven a significant improvement over the prior implementation).

Since this code has already been tested and provides the results we want, we can defer this change as follow-up work after the Code Freeze.